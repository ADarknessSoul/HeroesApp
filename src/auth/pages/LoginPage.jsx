import { useContext } from "react";
import { useNavigate } from "react-router-dom";
import { AuthContext } from "../context/AuthContext";

export const LoginPage = () => {

  const { Login } = useContext(AuthContext);

  const navigate = useNavigate();

  const onLogin = () => {

    const lastPath = JSON.parse(localStorage.getItem('lastPath')) || '/';

    Login('Usuario Gen√©rico');

    navigate(lastPath, {

        replace: true,

    });

  }
  
  return (
    <div className="form container">

      <h1 className="h1 text-center">Login</h1>

      {/* <h1>Login</h1>
      <hr />

      <button 
        className="btn btn-primary"
        onClick={ onLogin }
      >  
      
        Login

      </button> */}

      <form onSubmit={onLogin}>

        <div className="form-group">

          <label htmlFor="email-address">Email address</label>
          <input type="email" className="form-control mt-2" id="email-address" placeholder="someone@email.com"/>
          <small className="small text-muted">This field is not necessary for this demo</small>

        </div>

        <div className="form-group">

          <label htmlFor="user-password">Password</label>
          <input type="password" className="form-control mt-2" id="email-address" placeholder="someone@email.com"/>
          <small className="small text-muted">This field is not necessary for this demo</small>

        </div>
        
        <input type="submit" className="btn btn-outline-primary col-12 mt-3"/>

      </form>



    </div>
  )
}
