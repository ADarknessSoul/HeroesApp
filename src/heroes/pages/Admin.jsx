import { useContext, useState } from "react"
import { AuthContext } from "../../auth";
import { Navigate } from "react-router-dom";
import { addDoc, collection } from "firebase/firestore";
import { db } from '../../config/firebase';

export const Admin = () => {

    const { user } = useContext(AuthContext);

    const [createHeroId, setCreateHeroId] = useState("");
    const [createPublisher, setCreatePublisher] = useState("");
    const [createuserNames, setCreateUserNames] = useState("");
    const [createsuperHero, setCreateSuperHero] = useState("");
    const [createFirstAppearance, setCreateFirstAppearance] = useState("");
    const [createAlterEgo, setCreateAlterEgo] = useState("");
    const [createComicName, setCreateComicName] = useState("");
    const [createComicAuthor, setCreateComicAuthor] = useState("");
    const [createComicDate, setCreateComicDate] = useState(0);
    const [selectedImage, setSelectedImage] = useState(null);


    const heroesCollectionRef = collection(db, "heroes");

    const onSubmitHeroe = async() => {

        // const success = document.getElementById('success');

        const finalPublisher = `${createPublisher} Comics`;
        const finalHeroId = `${createPublisher.toLowerCase()}-${createsuperHero.toLowerCase().trim()}`;

        //Codigo para subir la imagen a un servidor de node y guardarlo en una carpeta (no funciona)
        // const newFileName = `${finalHeroId.toLowerCase().trim()}.jpg`;

        // const modifiedImage = new File([selectedImage], newFileName, {
        //   type: selectedImage.type,
        //   lastModified: selectedImage.lastModified,
        // });

        // const formData = new FormData();
        // formData.append('image', modifiedImage);

        // console.log(modifiedImage);

        // fetch('http://localhost:3000/upload', {
        //     method: 'POST',
        //     headers: {
        //         // Set the Content-Type header explicitly
        //         'Content-Type': 'multipart/form-data',
        //       },
        //     body: formData,
        //   })
        //     .then((response) => response.json())
        //     .then((data) => {
        //       console.log(data); // Response from the server
        //     })
        //     .catch((error) => {
        //       console.error('Error uploading image:', error);
        //     });


        try {

            await addDoc(heroesCollectionRef, {
                heroId: finalHeroId, 
                publisher: finalPublisher,
                character: {

                    alterEgo: createAlterEgo,
                    firstAppearance: createFirstAppearance,
                    superHero: createsuperHero,
                    userNames: createuserNames

                },
                bestComic: {

                    comicName: createComicName,
                    comicAuthor: createComicAuthor,
                    comicDate: createComicDate,

                }
                
            });

            success.classList.remove("d-none");

        } catch(err) {

            console.error(err);
            success.classList.add("d-none");

        }

        
    }

        //Aqui va el código Javascript


        const handleImageUpload = (event) => {
            setSelectedImage(event.target.files[0]);
          };

  return (
    <>

    {

        user.admin ? (

            <>
                   
        <h2 className= "text-center formulario_title ">Add a comic</h2>
        <div className="text-center formulario_container" >

            <form className= "formulario ">

            
            <div className="formulario_group">
                <h4 className="formulario_title formulario_labelA" > Publisher </h4>
            </div>

            <div className="form-check mb-3">
                <input className="form-check-input" type="radio" name="publishers" id="Marvel" onChange={(e) => setCreatePublisher(e.target.id)}/>
                <label className="form-check-label text-black" htmlFor="MarvelRadio">
                    Marvel
                </label>
            </div>

            <div className="form-check">
                <input className="form-check-input" type="radio" name="publishers" id="DC" onChange={(e) => setCreatePublisher(e.target.id)}/>
                <label className="form-check-label text-black" htmlFor="DCRadio">
                    DC
                </label>
            </div>              
              
              <div className="formulario_group">
                        <h4 className="text-center mt-5" > Data of heroes </h4>
              </div>

                <div className="formulario_group">
                   
                    <label htmlFor="identidad_secreta" className="formulario_labelA">Identidad Secreta:</label>
                    <input type="text" className="formulario_input"  id="identidad_secreta" name="identidad_secreta" required onChange={(e) => setCreateAlterEgo(e.target.value)}></input>
                </div>

                <div className="formulario_group">
                    <label htmlFor="primera_aparicion" className="formulario_labelA">Primera Aparición:</label>
                    <input type="text"  className="formulario_input" id="primera_aparicion" name="primera_aparicion" required onChange={(e) => setCreateFirstAppearance(e.target.value)}></input>
                </div>

                <div className="formulario_group">
                    <label htmlFor="nombre_heroe" className="formulario_labelA">Nombre de Héroe:</label>
                    <input type="text" className="formulario_input"  id="nombre_heroe" name="nombre_heroe" required onChange={(e) => setCreateSuperHero(e.target.value)}></input>
                </div>

                <div className="formulario_group">
                    <label htmlFor="anteriores_identidades" className="formulario_labelA" >Anteriores Identidades Secretas:</label>
                    <input type="text" className="formulario_input" id="anteriores_identidades" name="anteriores_identidades" required onChange={(e) => setCreateUserNames(e.target.value)}></input>
                </div>

                <div className="formulario_group">
                            <h4 className="formulario_title formulario_labelA" > Best Comic </h4>
                </div>

                 <div className="formulario_group">
                    <label htmlFor="autor" className="formulario_label">Autor:</label>
                    <input type="text" className= "formulario_input" id="autor" name="autor" required onChange={(e) => setCreateComicAuthor(e.target.value)}></input>
                 </div>
         
                
                <div className="formulario_group">
                    <label htmlFor="fecha_lanzamiento" className="formulario_labelA" >Fecha de Lanzamiento:</label>
                    <input type="number"  className="formulario_input" id="fecha_lanzamiento" name="fecha_lanzamiento" required onChange={(e) => setCreateComicDate(Number(e.target.value))}></input>
                </div>

                <div className="formulario_group">
                    <label htmlFor="titulo" className = "formulario_labelA" >Título del Comic:</label>
                    <input type="text" className="formulario_input" id="titulo" name="titulo" required onChange={(e) => setCreateComicName(e.target.value)}></input>
                </div>

                {/* <div className="formulario_group">
                    <label htmlFor="imagenHero" className = "formulario_labelA" >Imagen del comic:</label>
                    <input type="file" accept=".jpg" className="form-control mt-3" id="imagenHero" name="imagenHero" required onChange={handleImageUpload}></input>
                </div> */}

                <div className="modal-footer pt-4">
                     <input  type="button" className = "formulario-submit mx-auto w-50" value="Guardar" onClick={onSubmitHeroe}></input>
                </div>                
                    
            </form>

            <div id="success" className="mt-5 container text-center bg-success p-1 rounded d-none">

                <h2 className="text-white">Usuario creado con éxito</h2>

            </div>

        </div> 
            </>

        ) : (

            <Navigate to="/marvel"/>

        )

    }


    </>
  )
}
