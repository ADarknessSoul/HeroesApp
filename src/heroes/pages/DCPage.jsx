import { HeroList } from "../components"
import { db } from "../../config/firebase";
import { collection, onSnapshot, query, where } from 'firebase/firestore';
import { useEffect, useState } from "react";

export const DCPage = () => {
  
  const [heroesList, setHeroesList] = useState([]);
  const heroesCollectionRef = collection(db, "heroes");

  useEffect(() => {

    const q = query(heroesCollectionRef, where("publisher", "==", "DC Comics"));

    const unsubscribe = onSnapshot(q, (snapshot) => {

      const data = snapshot.docs.map((doc) => ({

        ...doc.data(),
        id: doc.id,

      }));

      setHeroesList(data);

    });

    return () => unsubscribe();

  }, []);

  return (
    <>
    
      <h1 className ="animate__animated animate__flash mt-4 h1">DC Comics</h1>
      <hr />

      <HeroList heroesList={heroesList}/>
    
    </>
  )
}
