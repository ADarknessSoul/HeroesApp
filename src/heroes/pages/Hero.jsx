import { useMemo } from "react";
import { Navigate, useNavigate, useParams } from "react-router-dom"
import { collection, onSnapshot, query, where, doc, updateDoc } from 'firebase/firestore';
import { db } from '../../config/firebase';
import { useEffect, useState } from "react";

export const Hero = () => {
  
  const [heroesList, setHeroesList] = useState([]);
  const heroesCollectionRef = collection(db, "heroes");

  const [updatedComicName, setUpdatedComicName] = useState("");
  const [updatedComicAuthor, setUpdatedComicAuthor] = useState("");
  const [updatedComicDate, setUpdatedComicDate] = useState(0);

  const { id } = useParams();

  useEffect(() => {

    const q = query(heroesCollectionRef, where("heroId", "==", id.toString()));

    const unsubscribe = onSnapshot(q, (snapshot) => {

      const data = snapshot.docs.map((doc) => ({

        ...doc.data(),
        id: doc.id,

      }));

      setHeroesList(data);

    });

    return () => unsubscribe();

  }, []);
  
  const navigate = useNavigate();

  const onNavigate = () => {

    navigate(-1);

  }

  const updateHeroeTitle = async(id) => {

    const heroesDoc = doc(db, "heroes", id);
    await updateDoc(heroesDoc, { "bestComic.comicName": updatedComicName});

  }

  const updateHeroeAuthor = async(id) => {

    const heroesDoc = doc(db, "heroes", id);
    await updateDoc(heroesDoc, { "bestComic.comicAuthor": updatedComicAuthor});

  }

  const updateHeroeDate = async(id) => {

    const heroesDoc = doc(db, "heroes", id);
    await updateDoc(heroesDoc, { "bestComic.comicDate": updatedComicDate.toString()});

  }

  return (
    <>
      { heroesList.length === 0 ? (
      <></>
    ) : (
      <>
        <div className="overlay">

          <div className="overlay-image animate__animated animate__fadeInLeft">

            <img
              src={`/heroes/${heroesList[0]?.heroId}.jpg`}
              alt={heroesList[0]?.character.superhero}
              className="rounded"
            />

          </div>

        </div>

        <div className="text-center mt-2 border mb-3 semi-rounded">

          <h2 className="h2 text-uppercase">{heroesList[0]?.character.superHero}</h2>
          <hr />

          <p>Alter ego: {heroesList[0]?.character.alterEgo}</p>
          <p>Publisher: {heroesList[0]?.publisher}</p>
          <p>First appereance: {heroesList[0]?.character.firstAppearance}</p>
          <hr />

          <h3 className="h3 text-uppercase">Characters</h3>
          <hr />

          <p>{heroesList[0]?.character.userNames}</p>

          <hr />

          <h2 className="h1 text-success">Best comic</h2>
          <hr />

          <div>

            <img src="/public/heroes/trophy.jpg" alt="trophy" width="300px" height="300px"/>

          </div>

          <div className="d-flex mb-4 flex-column align-items-start p-3">

            <h3 className="h3 text-muted" style={{margin: "0 auto"}}>Comic: <span className="h4 text-secondary">{heroesList[0].bestComic.comicName}</span></h3>

            <div className="container border mt-3 p-3 mb-2">

              <label htmlFor="updateComicName">Nombre del comic:</label>
              <input type="text" className="form-control" onChange={(e) => setUpdatedComicName(e.target.value)}/>
              <button className="btn btn-warning col-12 mt-2" onClick={() => updateHeroeTitle(heroesList[0].id)}>Actualizar</button>

            </div>
            
            <h3 className="h3 text-muted" style={{margin: "0 auto"}}>Author: <span className="h4 text-secondary">{heroesList[0].bestComic.comicAuthor}</span></h3>

            
            <div className="container border mt-3 p-3 mb-2">

              <label htmlFor="updateComicAuthor">Nombre del autor:</label>
              <input type="text" className="form-control" onChange={(e) => setUpdatedComicAuthor(e.target.value)}/>
              <button className="btn btn-warning col-12 mt-2" onClick={() => updateHeroeAuthor(heroesList[0].id)}>Actualizar</button>

            </div>

            <h3 className="h3 text-muted" style={{margin: "0 auto"}}>Release date: <span className="h4 text-secondary">{heroesList[0].bestComic.comicDate}</span></h3>

            <div className="container border mt-3 p-3 mb-2">

              <label htmlFor="updateComicDate">Fecha de estreno:</label>
              <input type="number" className="form-control" onChange={(e) => setUpdatedComicDate(Number(e.target.value))}/>
              <button className="btn btn-warning col-12 mt-2" onClick={() => updateHeroeDate(heroesList[0].id)}>Actualizar</button>

            </div>

          </div>

          <button onClick={onNavigate} className="btn btn-primary w-100">
            Return
          </button>

        </div>
      </>
    )}
    
    </>
  )
}

