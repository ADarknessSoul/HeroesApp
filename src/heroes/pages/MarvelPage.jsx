import { HeroList } from "../components"

export const MarvelPage = () => {

  return (

    <>
      
      <h1 className="animate__animated animate__flash mt-4 h1">Marvel Comics</h1>
      <hr />

      <HeroList publisher={'Marvel Comics'}/>
    
    </>

  )
}
