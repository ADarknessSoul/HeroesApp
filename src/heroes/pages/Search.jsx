import { useLocation, useNavigate } from "react-router-dom";
import { useForm } from "../../hooks/useForm";
import queryString from "query-string";
import { HeroCard } from "../components/HeroCard";
// import { getHeroesByName } from "../helpers";
import { db } from "../../config/firebase";
import { collection, onSnapshot, query, where } from 'firebase/firestore';
import { useEffect, useState } from "react";

export const Search = () => {

  const navigate = useNavigate();
  const location = useLocation();

  const [heroesList, setHeroesList] = useState([]);
  const heroesCollectionRef = collection(db, "heroes");

  const { q  = '' } = queryString.parse(location.search);

  useEffect(() => {

    console.log(q);
    const newq = query(heroesCollectionRef, where("character.superHero", "==", q));

    const unsubscribe = onSnapshot(newq, (snapshot) => {

      const data = snapshot.docs.map((doc) => ({

        ...doc.data(),
        id: doc.id,

      }));

      setHeroesList(data);

    });

    return () => unsubscribe();

  }, [q]);

  //const heroes = getHeroesByName(q);
  const showSearch = (q.length === 0);
  // const showError = (q.length > 0) && heroes.length === 0;
  // const showError = (q.length > 0);

  const { searchText, onInputChange } = useForm({

    searchText: q,
  
  });
  
  const onSearchSubmit = (e) => {
    
    e.preventDefault();
    // if( searchText.trim().length <= 0) return;
    navigate(`?q=${searchText}`);
  
  }

  return (
    <>
    
      <h1 className="text-center text-uppercase">Heroes database</h1>
      <hr />

      <div className="row">

        <div className="col-sm-5">

          <h2 className="h2">Searching</h2>
          <hr />
          <form onSubmit={ onSearchSubmit } data-testid="formulario">

            <input 
              type="text"
              placeholder="Example: Wonder Woman"
              className="form-control"
              name="searchText" 
              autoComplete="off"
              value={ searchText }
              onChange={ onInputChange }
            />

            <button className="btn btn-outline-primary mt-2">Search</button>

          </form>

        </div> {/*.col-5 */}

          <div className="col-sm-7">

          <h2>Results</h2>
          <hr />

          {/* {

            (q === '')
            ?          
            <div className="alert alert-primary">

            Search a hero

            </div>
            : (heroes.length === 0) &&
            <div className="alert alert-danger">

            The hero <b>{q}</b> was not found

            </div>

          } */}

          <div data-testid="SearchaHero" className="alert alert-primary" style={{display: showSearch ? "" : "none"}}>

          Search a hero

          </div>

          <div data-testid="errorHero" className="alert alert-danger" style={{display: (q.length > 0) && heroesList.length === 0 ? "" : "none"}}>

            The hero <b>{q}</b> was not found

          </div>

          {

            heroesList.map( Hero => (

              <HeroCard key={Hero.heroId} Hero={Hero} withoutButton={true}/>

            ))

          }

          {

            console.log(heroesList)

          }


          {/* <HeroCard /> */}

        </div> {/*.col-7 */}

      </div> {/*.row */}

    </>
  )
}
