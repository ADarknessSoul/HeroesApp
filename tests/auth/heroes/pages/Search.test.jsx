import { fireEvent, render, screen } from "@testing-library/react";
import { Search } from "../../../../src/heroes/pages/Search";
import { MemoryRouter } from "react-router-dom";

const mockedUseNavigate = jest.fn();

jest.mock('react-router-dom', () => ({  //Se hace un mock completo de la libreria, pero solo sobrescribimos la funciÃ³n que necesitamos

    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockedUseNavigate

}))


describe('Pruebas en <Search/>', ()=> {

    beforeEach(() => jest.clearAllMocks());

    test('Debe de mostrarse correctamente con valores por defecto', ()=> {

        const {container} = render(

            <MemoryRouter>

                <Search/>

            </MemoryRouter>

        );

        expect( container ).toMatchSnapshot();

    }); 


    test('Debe mostrar a batman y el input con el valor del queryString', ()=> {

        render(

            <MemoryRouter initialEntries={['/search?q=batman']}>

                <Search/>

            </MemoryRouter>

        );

        const inputValue = screen.getByRole('textbox');

        const image = screen.getByRole('img');

        const SearchaHero = screen.getByTestId('SearchaHero');

        expect(SearchaHero.style.display).toBe('none');

        expect(image.src).toContain('/heroes/dc-batman.jpg');

        expect(inputValue.value).toBe('batman');

    }); 

    test('Debe mostrar un error si no se encuentra el heroe (batman123)', ()=> {
    
        render(

            <MemoryRouter initialEntries={['/search?q=batman123']}>

                <Search/>

            </MemoryRouter>

        );

        const error = screen.getByTestId('errorHero');

        expect(error.style.display).not.toBe("none");
    
    });

    test('Debe llamar el navigate a la pantalla nueva', ()=> {
    
        render(

            <MemoryRouter initialEntries={['/search']}>

                <Search/>

            </MemoryRouter>

        );

        const input = screen.getByRole('textbox');

        fireEvent.change( input, {target: { name: 'searchText', value: 'superman'}});

        const form = screen.getByTestId('formulario');

        fireEvent.submit(form);

        expect( mockedUseNavigate ).toHaveBeenCalledWith('?q=superman');
    
    });

});