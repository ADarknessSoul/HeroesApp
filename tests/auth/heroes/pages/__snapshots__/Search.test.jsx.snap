// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <Search/> Debe de mostrarse correctamente con valores por defecto 1`] = `
<div>
  <h1
    className="text-center text-uppercase"
  >
    Heroes database
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-sm-5"
    >
      <h2
        className="h2"
      >
        Searching
      </h2>
      <hr />
      <form
        data-testid="formulario"
      >
        <input
          autocomplete="off"
          className="form-control"
          name="searchText"
          placeholder="Example: Wonder Woman"
          type="text"
          value=""
        />
        <button
          className="btn btn-outline-primary mt-2"
        >
          Search
        </button>
      </form>
    </div>
     
    <div
      className="col-sm-7"
    >
      <h2>
        Results
      </h2>
      <hr />
      <div
        className="alert alert-primary"
        data-testid="SearchaHero"
      >
        Search a hero
      </div>
      <div
        className="alert alert-danger"
        data-testid="errorHero"
        style="display: none;"
      >
        The hero 
        <b />
         was not found
      </div>
    </div>
     
  </div>
   
</div>
`;
